sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,236 @@
+
+        private readonly IOperation<TResult> _operation;
+
+        private TResult _value;
+
+        public OperationInternal(ClientDiagnostics clientDiagnostics, IOperation<TResult> operation) : this(clientDiagnostics)
+        {
+            _operation = operation;
+            OperationTypeName = operation.GetType().Name;
+        }
+
+        protected OperationInternal(ClientDiagnostics clientDiagnostics)
+        {
+            ClientDiagnostics = clientDiagnostics;
+            DefaultPollingInterval = TimeSpan.FromSeconds(1);
+            ScopeAttributes = new Dictionary<string, string>();
+        }
+
+        public IDictionary<string, string> ScopeAttributes { get; }
+
+        public bool HasValue { get; private set; }
+
+        public bool HasCompleted { get; set; }
+
+        public TResult Value
+        {
+            get
+            {
+                if (HasValue)
+                {
+                    return _value;
+                }
+                else if (OperationFailedException != null)
+                {
+                    throw OperationFailedException;
+                }
+                else
+                {
+                    throw new InvalidOperationException("The operation has not completed yet.");
+                }
+            }
+            set
+            {
+                if (value is null)
+                {
+                    throw new ArgumentNullException(nameof(Value));
+                }
+
+                _value = value;
+                HasValue = true;
+            }
+        }
+
+        public Response RawResponse { get; set; }
+
+        public string OperationTypeName { get; set; }
+
+        public TimeSpan DefaultPollingInterval { get; set; }
+
+        protected ClientDiagnostics ClientDiagnostics { get; }
+
+        protected RequestFailedException OperationFailedException { get; set; }
+
+        public async ValueTask<Response> UpdateStatusAsync(CancellationToken cancellationToken) =>
+            await UpdateStatusAsync(async: true, cancellationToken).ConfigureAwait(false);
+
+        public Response UpdateStatus(CancellationToken cancellationToken) =>
+            UpdateStatusAsync(async: false, cancellationToken).EnsureCompleted();
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                Response response = await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
+
+                if (HasCompleted)
+                {
+                    return Response.FromValue(Value, response);
+                }
+
+                var pollingInterval = response.Headers.TryGetValue(RetryAfterHeaderName, out string retryAfterValue)
+                    && int.TryParse(retryAfterValue, out int delayInSeconds)
+                    ? TimeSpan.FromSeconds(delayInSeconds) : DefaultPollingInterval;
```

Comment:
Wondering what's the right approach here. This is what we're doing:
- If `pollingInterval` is specified: use it, regardless of the `Retry-After` header.
- If `pollingInterval` is not specified: use the `Retry-After` header. If we can't find it, use the `DefaultPollingInterval`.

Regarding the first scenario, ignoring the `Retry-After` header seems right to me. The user has clearly specified the polling interval to use, so I think we should not do any extra logic.

Reply:
>Regarding the first scenario, ignoring the Retry-After header seems right to me. The user has clearly specified the polling interval to use, so I think we should not do any extra logic.

That's one way to look at it. The other is that the user might be querying to often and the service know that there would be no result for some time. In the pipeline retry policy we prefer the server-provided delay if it's longer. (https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs#L195-L196)

Reply:
Hmmm. I see. Choosing the longer one seems wise. One concern I had was in case the user specified the polling interval to limit the amount of requests to the service, but this won't be a problem if we take the longer one. I think it's worth keeping it consistent with the retry policy behavior.